### ISSUE TITLE
External Svelte component (from pkg.svelte) leads to code duplication and breaking of setContext

### ISSUE THREAD

**Issue by flekschas** [rollup/rollup-plugin-svelte#55]
I hope this is the correct repo to file an issue.

I am having a Svelte component that I install via npm (`npm link ...` for testing purposes if that matters) and that points to the uncompiled source via `pkg.svelte`. This component internally uses Svelte's `setContext` to establish a context. Unfortunately, the context is not available in my main Svelte app, which is importing the component. After some debugging, I found out that `setContext` sets the context on the current Svelte instance and it turns out that the imported component and the _importing_ component are initiated from different Svelte instances.

Here's a small example:

```html
<!-- App.svelte -->
<script>
  import ExternalComponent from 'external-component';  // external
  import Test from './Test.svelte';  // internal
</script>

<ExternalComponent><Test /><ExternalComponent>


<!-- Test.svelte -->
<script>
  import { getContext } from 'svelte';
  const { hi } = getContext('external-component');  // This line causes an error because `getContext('external-component')` is not available!
</script>

<p>{hi}</p>


<!-- ExternalComponent.svelte that is imported as 'external-component' -->
<script>
  import { setContext } from 'svelte';
  setContext('external-component', { hi: 'Hi there!' });
</script>
```

In general, the bundling works because if I pass `setContent` into `ExternalComponent` everything works as expected. E.g., the following modified code will render `<p>Hi there!</p>`:

```html
<!-- App.svelte -->
<script>
  import { setContext } from 'svelte';
  import ExternalComponent from 'external-component';  // external
  import Test from './Test.svelte';  // internal
</script>

<ExternalComponent {setContext}><Test /><ExternalComponent>


<!-- ExternalComponent.svelte that is imported as 'external-component' -->
<script>
  import { setContext: baseSetContext } from 'svelte';

  export let setContext = baseSetContext;

  setContext('external-component', { hi: 'Hi there!' });
</script>
```

The workaround with passing down `setContext` works but it'd be great if this could be avoided. I assume that somehow `get_current_component()` returns a different component for the internal and external Svelte component. Is there a way to inspect the component's context object directly?

I assume this is an issue with the bundling because the identical code worked just fine as long as `ExternalComponent.svelte` was part of the other app.

**Update:** Not sure if this is too relevant/surprising but I get `console.log(setContext === baseSetContext) // => false` in the modified `ExternalComponent.svelte`, which to me sounds like somehow code is duplicated or there are two instances of Svelte running. But I am missing a deep background of what the Svelte compiler is really doing to be sure.

**Comment by flekschas** [comment]
After looking at the generated code I think I know what the issue is: rollup somehow generates two copies of Svelte. E.g., the output bundle contains `SvelteComponent` and `SvelteComponent$1`.

All _internal_ components extend `SvelteComponent` while the external component extends `SvelteComponent$1`. Hence, `setContext` as imported by the external component registers a context on `SvelteComponent$1` while internal components try to get the context from `SvelteComponent`, which obviously does not work.

Besides the issue with the context, this blows up the file size of the bundle too.

I am using
- rollup: v1.11.3
- rollup-plugin-commonjs: v9.3.4
- rollup-plugin-node-resolve: v4.2.4
- rollup-plugin-svelte: v5.0.3
- svelte: v3.3.0

**Comment by mrkishi** [comment]
What Svelte version does the external package depend on? What package manager are you using?

Actually, seeing that you have the external component via `npm link`, I'd expect it to have its own `node_modules` and thus its own Svelte copy. I believe this wouldn't happen when you `npm install other-package` unless you're using a package manager like `pnpm` or if the versions are incompatible.

**Comment by flekschas** [comment]
Both use the exact same versions of all packages. I use npm as the package manager.

I'll try to see if I can minic `npm i external-package`. It'd make sense if it's the other package's Svelte instance.

**Comment by mrkishi** [comment]
If you just delete `svelte` from the `external-package`'s `node_modules` it should work. It should probably have svelte as a peer dependency instead of a dependency.

**Comment by flekschas** [comment]
The compilation won't work though. I initially didn't have Svelte installed but it needs to be there to get rollup-plugin-svelte to work for bundling. I could just not bundle the external package since it's anyway only useful for Svelte but removing `svelte` from `node_module `also breaks the import by the bundler of the other app. Rollup of the _internal_ app is not able to resolve the dependencies to `svelte` coming from `import { setContext } from 'svelte';` in the external package.

Essentially I am seeing:

```bash
(!) Unresolved dependencies
https://rollupjs.org/guide/en#warning-treating-module-as-external-dependency
svelte (imported by ../svelte-simple-modal/src/Modal.svelte)
svelte/internal (imported by ../svelte-simple-modal/src/Modal.svelte)
(!) Missing global variable names
Use output.globals to specify browser global variable names corresponding to external modules
svelte/internal (guessing 'internal')
svelte (guessing 'svelte')```

I assumed that with `pkg.svelte` imports from the core svelte package are merged but they seem to be treated as two different packages.

**Comment by flekschas** [comment]
I think I found the solution. 🎉 

I had to add `dedupe: ['svelte', 'svelte/internal']` to my internal svelte app's `rollup.config.js` to ensure that the _root_ `node_modules` are used to resolve `svelte`.

Thanks for your help @mrkishi. I'll close the issue. It might still be worth thinking about setting the base svelte packages in `dedupe` when an external Svelte package is imported in its uncompiled form via `pkg.svelte` as I don't think users are expected to use different svelte version but maybe this is really only an issue when using `setContext`. 🤷‍♂️
